File: client/lib/models/game.dart
/// Game data models matching the server API responses.

class GameVersion {
  final int id;
  final String platform;
  final String filename;
  final String filePath;
  final int fileSize;
  final String sourceType;
  final int? sourceId;
  final String? sourcePath;
  final String? extractPassword;

  GameVersion({
    required this.id,
    required this.platform,
    required this.filename,
    required this.filePath,
    required this.fileSize,
    this.sourceType = "local",
    this.sourceId,
    this.sourcePath,
    this.extractPassword,
  });

  factory GameVersion.fromJson(Map<String, dynamic> json) {
    return GameVersion(
      id: json["id"] ?? 0,
      platform: json["platform"] ?? "",
      filename: json["filename"] ?? "",
      filePath: json["file_path"] ?? "",
      fileSize: json["file_size"] ?? 0,
      sourceType: json["source_type"] ?? "local",
      sourceId: json["source_id"],
      sourcePath: json["source_path"],
      extractPassword: json["extract_password"],
    );
  }
}

class Tag {
  final int id;
  final String name;
  final String color;
  final String source;
  final double weight;
  final bool isSpoiler;

  Tag({
    required this.id,
    required this.name,
    required this.color,
    this.source = "",
    this.weight = 0,
    this.isSpoiler = false,
  });

  factory Tag.fromJson(Map<String, dynamic> json) {
    return Tag(
      id: json["id"] ?? 0,
      name: json["name"] ?? "",
      color: json["color"] ?? "#3B82F6",
      source: json["source"]?.toString() ?? "",
      weight: _readDouble(json["weight"] ?? json["rating"]),
      isSpoiler: _readBool(json["is_spoiler"] ?? json["spoiler"]),
    );
  }

  static double _readDouble(dynamic value) {
    if (value is num) return value.toDouble();
    if (value is String) return double.tryParse(value) ?? 0;
    return 0;
  }

  static bool _readBool(dynamic value) {
    if (value is bool) return value;
    if (value is num) return value != 0;
    if (value is String) {
      final normalized = value.trim().toLowerCase();
      return normalized == "true" || normalized == "1" || normalized == "yes";
    }
    return false;
  }
}

class GameSummary {
  final int id;
  final String name;
  final String? companyName;
  final String? developer;
  final String folderPath;
  final String? coverPath;
  final bool isNsfw;
  final String platformSummary;
  final List<String> tagNames;
  final String importedAt;
  final int length;
  final int lengthMinutes;

  GameSummary({
    required this.id,
    required this.name,
    this.companyName,
    this.developer,
    required this.folderPath,
    this.coverPath,
    this.isNsfw = false,
    required this.platformSummary,
    required this.tagNames,
    required this.importedAt,
    this.length = 0,
    this.lengthMinutes = 0,
  });

  factory GameSummary.fromJson(Map<String, dynamic> json) {
    return GameSummary(
      id: json["id"] ?? 0,
      name: json["name"] ?? "",
      companyName: json["company_name"],
      developer: json["developer"],
      folderPath: json["folder_path"] ?? "",
      coverPath: json["cover_path"],
      isNsfw: json["is_nsfw"] == true,
      platformSummary: json["platform_summary"] ?? "",
      tagNames: List<String>.from(json["tag_names"] ?? []),
      importedAt: json["imported_at"] ?? "",
      length: json["length"] ?? 0,
      lengthMinutes: json["length_minutes"] ?? 0,
    );
  }
}

class GameDetail {
  final int id;
  final String name;
  final String? companyName;
  final int rootId;
  final String folderPath;
  final String? coverPath;
  final String? bgPath;
  final bool isNsfw;
  final String? developer;
  final String? description;
  final String? releaseDate;
  final String? vndbId;
  final String? steamId;
  final String? bangumiId;
  final String? hikarinagiId;
  final int length;
  final int lengthMinutes;
  final bool isDeleted;
  final String importedAt;
  final String updatedAt;
  final List<GameVersion> versions;
  final List<Tag> tags;

  GameDetail({
    required this.id,
    required this.name,
    this.companyName,
    required this.rootId,
    required this.folderPath,
    this.coverPath,
    this.bgPath,
    this.isNsfw = false,
    this.developer,
    this.description,
    this.releaseDate,
    this.vndbId,
    this.steamId,
    this.bangumiId,
    this.hikarinagiId,
    this.length = 0,
    this.lengthMinutes = 0,
    required this.isDeleted,
    required this.importedAt,
    required this.updatedAt,
    required this.versions,
    required this.tags,
  });

  factory GameDetail.fromJson(Map<String, dynamic> json) {
    return GameDetail(
      id: json["id"] ?? 0,
      name: json["name"] ?? "",
      companyName: json["company_name"],
      rootId: json["root_id"] ?? 0,
      folderPath: json["folder_path"] ?? "",
      coverPath: json["cover_path"],
      bgPath: json["bg_path"],
      isNsfw: json["is_nsfw"] == true,
      developer: json["developer"],
      description: json["description"],
      releaseDate: json["release_date"],
      vndbId: json["vndb_id"],
      steamId: json["steam_id"],
      bangumiId: json["bangumi_id"],
      hikarinagiId: json["hikarinagi_id"],
      length: json["length"] ?? 0,
      lengthMinutes: json["length_minutes"] ?? 0,
      isDeleted: json["is_deleted"] ?? false,
      importedAt: json["imported_at"] ?? "",
      updatedAt: json["updated_at"] ?? "",
      versions: (json["versions"] as List<dynamic>?)
              ?.map((v) => GameVersion.fromJson(v as Map<String, dynamic>))
              .toList() ??
          [],
      tags: (json["tags"] as List<dynamic>?)
              ?.map((t) => Tag.fromJson(t as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}

File: server/models/game.py
"""Game and GameVersion models."""

from __future__ import annotations

import enum
from datetime import datetime

from sqlalchemy import (
    Column,
    DateTime,
    Enum,
    ForeignKey,
    Integer,
    String,
    Text,
    Boolean,
    BigInteger,
)
from sqlalchemy.orm import relationship

from database import Base


class Platform(str, enum.Enum):
    PC = "PC"
    KRKR = "KRKR"
    TYRANOR = "Ty"
    ONS = "ONS"
    DIRECT = "直装"


class Company(Base):
    __tablename__ = "companies"

    id = Column(Integer, primary_key=True, autoincrement=True)
    name = Column(String(255), nullable=False, unique=True)
    created_at = Column(DateTime, default=datetime.utcnow)

    games = relationship("Game", back_populates="company")


class Game(Base):
    __tablename__ = "games"

    id = Column(Integer, primary_key=True, autoincrement=True)
    name = Column(String(512), nullable=False)
    company_id = Column(Integer, ForeignKey("companies.id"), nullable=True)
    root_id = Column(Integer, ForeignKey("root_directories.id"), nullable=False)
    folder_path = Column(String(1024), nullable=False, unique=True)

    # Metadata (populated by scrapers in Phase 2)
    cover_path = Column(String(1024), nullable=True)
    bg_path = Column(String(1024), nullable=True)
    developer = Column(String(512), nullable=True)
    description = Column(Text, nullable=True)
    release_date = Column(String(64), nullable=True)
    vndb_id = Column(String(32), nullable=True)
    steam_id = Column(String(32), nullable=True)
    bangumi_id = Column(String(32), nullable=True)

    length = Column(Integer, default=0)
    length_minutes = Column(Integer, default=0)

    is_deleted = Column(Boolean, default=False)
    imported_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    company = relationship("Company", back_populates="games")
    versions = relationship("GameVersion", back_populates="game", cascade="all, delete-orphan")
    tags = relationship("GameTag", back_populates="game", cascade="all, delete-orphan")


class GameVersion(Base):
    __tablename__ = "game_versions"

    id = Column(Integer, primary_key=True, autoincrement=True)
    game_id = Column(Integer, ForeignKey("games.id"), nullable=False)
    platform = Column(Enum(Platform), nullable=False)
    filename = Column(String(512), nullable=False)
    file_path = Column(String(1024), nullable=False)
    source_type = Column(String(32), nullable=False, default="local")
    source_id = Column(Integer, nullable=True)
    source_path = Column(String(1024), nullable=True)
    file_size = Column(BigInteger, default=0)
    extract_password = Column(String(256), nullable=True)

    game = relationship("Game", back_populates="versions")


class GameTag(Base):
    __tablename__ = "game_tags"

    id = Column(Integer, primary_key=True, autoincrement=True)
    game_id = Column(Integer, ForeignKey("games.id"), nullable=False)
    tag_id = Column(Integer, ForeignKey("tags.id"), nullable=False)

    game = relationship("Game", back_populates="tags")
    tag = relationship("Tag", back_populates="games")
